### ISSUE TITLE
How to reproduce the result on the Bi-encoder model for dstc7

### ISSUE THREAD

**Issue by huaxiadadi** [facebookresearch/ParlAI#3133]
Hi Kurt @klshuster , could you tell me the hyperparameters for reproducing the reult on the Bi-encoder model for dstc7? 
And btw, how do you get the utterances input token ids  for each example for the utterance encoder? 
Thanks!

**Comment by stephenroller** [comment]
Sounds like we should add some notes in the project README, since this has come up twice.

**Comment by huaxiadadi** [comment]
> Sounds like we should add some notes in the project README, since this has come up twice.

Sounds good. It'll help a lot if you add some notes in the README.

**Comment by klshuster** [comment]
Apologies for the late response, this got buried...

Interestingly, we have had questions about dstc7 for the cross-encoder and the poly-encoder, but not yet for the bi-encoder!

The setup is very close to that of the poly-encoder; please use the following hyperparameters when running `parlai train_model`:

```
-t dstc7:DSTC7TeacherAugmentedSampled
--init-model zoo:pretrained_transformers/bi_model_huge_reddit/model  # change this if you want to use wikito instead
--batchsize 256  
--model transformer/biencoder 
--warmup_updates 100 
--lr-scheduler-patience 0 
--lr-scheduler-decay 0.4 
-lr 5e-05 
--data-parallel True 
--history-size 20 
--label-truncate 72 
--text-truncate 360 
-vp 5 
-veps 0.5 
--validation-metric accuracy 
--validation-metric-mode max 
--save-after-valid True 
--log_every_n_secs 20 
--candidates batch 
--dict-tokenizer bpe  
--dict-lower True 
--optimizer adamax 
--output-scaling 0.06 
--variant xlm 
--reduction_type mean 
--share-encoders False 
--learn-positional-embeddings True 
--n-layers 12 
--n-heads 12 
--ffn-size 3072 
--attention-dropout 0.1 
--relu-dropout 0.0 
--dropout 0.1 
--n-positions 1024 
--embedding-size 768 
--activation gelu 
--embeddings-scale False 
--n-segments 2 
--learn-embeddings True 
--share-word-embeddings False 
--dict-endtoken __start__ 
--fp16 True 
```
 I am putting out a PR shortly to add these instructions to the README.

As for tokenization - setting the `--dict-tokenizer bpe` flag will utilize the `SubwordBPE` tokenizer found in [this file](https://github.com/facebookresearch/ParlAI/blob/master/parlai/utils/bpe.py#L270). Text is converted to tokens via the `encode` function, and tokens to text via `decode`. Hope that answers your question

**Comment by huaxiadadi** [comment]
> Apologies for the late response, this got buried...
> 
> Interestingly, we have had questions about dstc7 for the cross-encoder and the poly-encoder, but not yet for the bi-encoder!
> 
> The setup is very close to that of the poly-encoder; please use the following hyperparameters when running `parlai train_model`:
> 
> ```
> -t dstc7:DSTC7TeacherAugmentedSampled
> --init-model zoo:pretrained_transformers/bi_model_huge_reddit/model  # change this if you want to use wikito instead
> --batchsize 256  
> --model transformer/biencoder 
> --warmup_updates 100 
> --lr-scheduler-patience 0 
> --lr-scheduler-decay 0.4 
> -lr 5e-05 
> --data-parallel True 
> --history-size 20 
> --label-truncate 72 
> --text-truncate 360 
> -vp 5 
> -veps 0.5 
> --validation-metric accuracy 
> --validation-metric-mode max 
> --save-after-valid True 
> --log_every_n_secs 20 
> --candidates batch 
> --dict-tokenizer bpe  
> --dict-lower True 
> --optimizer adamax 
> --output-scaling 0.06 
> --variant xlm 
> --reduction_type mean 
> --share-encoders False 
> --learn-positional-embeddings True 
> --n-layers 12 
> --n-heads 12 
> --ffn-size 3072 
> --attention-dropout 0.1 
> --relu-dropout 0.0 
> --dropout 0.1 
> --n-positions 1024 
> --embedding-size 768 
> --activation gelu 
> --embeddings-scale False 
> --n-segments 2 
> --learn-embeddings True 
> --share-word-embeddings False 
> --dict-endtoken __start__ 
> --fp16 True 
> ```
> 
> I am putting out a PR shortly to add these instructions to the README.
> 
> As for tokenization - setting the `--dict-tokenizer bpe` flag will utilize the `SubwordBPE` tokenizer found in [this file](https://github.com/facebookresearch/ParlAI/blob/master/parlai/utils/bpe.py#L270). Text is converted to tokens via the `encode` function, and tokens to text via `decode`. Hope that answers your question

Got it, thanks!

**Comment by klshuster** [comment]
going to close for now, please feel free to reopen if you have additional questions

**Comment by huaxiadadi** [comment]
Hi Kurt @klshuster , you've shown me the SubwordBPE for text tokenization but I want to know more about how did you process the data from the original 'ubuntu_train_subtask_1_sampled.json' file to the inputs(token ids and attention masks) of transformer.   I'm just curious if you add some sepcial person tokens to indicate different participants when encoding the contexts and how did you covert tokens to token ids. Could you give me  those codes that: 1. process the original dstc json file 2.tokenize context texts and get the tokens up to a given max length 3. convert those tokenized tokens to the input ids and input attention masks  for transformer like Bert.  Thanks!

**Comment by klshuster** [comment]
I will preface my answer by noting that providing an in depth explanation about how everything works in ParlAI would not be suitable as an answer to this question; however, providing _exact_ code pointers might be meaningless unless you have used ParlAI before and understand on a general level how everything is connected.

1. the dstc7 teacher is in [this file](https://github.com/facebookresearch/ParlAI/blob/master/parlai/tasks/dstc7/agents.py). specifically, the file you mentioned is [loaded here](https://github.com/facebookresearch/ParlAI/blob/master/parlai/tasks/dstc7/agents.py#L47) and examples are presented to agent [as in here](https://github.com/facebookresearch/ParlAI/blob/master/parlai/tasks/dstc7/agents.py#L76) 
2. examples are tokenized when the agent [observes them](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/torch_agent.py#L1644); when an agent observes, it [updates its history](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/torch_agent.py#L266); we didn't add person tokens for these agents, however one could by setting the appropriate flag (and [notice how they are set here](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/torch_agent.py#L286)). The history class [`parse`s the text](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/torch_agent.py#L222), which calls the dictionary's[ `vec2txt` function](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/dict.py#L731), where `bpe.decode` is called (subword BPE decode call is [here](https://github.com/facebookresearch/ParlAI/blob/master/parlai/utils/bpe.py#L322)).
3. all of our code using transformers is in [this file](https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/transformer/modules.py) (we have implemented our own transformer modules). Here you'll find the [`TransformerEncoder`](https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/transformer/modules.py#L333) which is used by the bi-encoder for its context and candidate encoder. The [`forward` function](https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/transformer/modules.py#L569) is divided into three steps; `forward_embedding`, `forward_layers`, and `reduce_output`. In [`forward_embedding`](https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/transformer/modules.py#L472) you'll find where we send the tokens through the embedding layers and compute the masks. If you want to see how attention masks are computed specifically, take a look at some of the modules in that file.

Hope that answers your questions.

**Comment by huaxiadadi** [comment]
> I will preface my answer by noting that providing an in depth explanation about how everything works in ParlAI would not be suitable as an answer to this question; however, providing _exact_ code pointers might be meaningless unless you have used ParlAI before and understand on a general level how everything is connected.
> 
> 1. the dstc7 teacher is in [this file](https://github.com/facebookresearch/ParlAI/blob/master/parlai/tasks/dstc7/agents.py). specifically, the file you mentioned is [loaded here](https://github.com/facebookresearch/ParlAI/blob/master/parlai/tasks/dstc7/agents.py#L47) and examples are presented to agent [as in here](https://github.com/facebookresearch/ParlAI/blob/master/parlai/tasks/dstc7/agents.py#L76)
> 2. examples are tokenized when the agent [observes them](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/torch_agent.py#L1644); when an agent observes, it [updates its history](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/torch_agent.py#L266); we didn't add person tokens for these agents, however one could by setting the appropriate flag (and [notice how they are set here](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/torch_agent.py#L286)). The history class [`parse`s the text](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/torch_agent.py#L222), which calls the dictionary's[ `vec2txt` function](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/dict.py#L731), where `bpe.decode` is called (subword BPE decode call is [here](https://github.com/facebookresearch/ParlAI/blob/master/parlai/utils/bpe.py#L322)).
> 3. all of our code using transformers is in [this file](https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/transformer/modules.py) (we have implemented our own transformer modules). Here you'll find the [`TransformerEncoder`](https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/transformer/modules.py#L333) which is used by the bi-encoder for its context and candidate encoder. The [`forward` function](https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/transformer/modules.py#L569) is divided into three steps; `forward_embedding`, `forward_layers`, and `reduce_output`. In [`forward_embedding`](https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/transformer/modules.py#L472) you'll find where we send the tokens through the embedding layers and compute the masks. If you want to see how attention masks are computed specifically, take a look at some of the modules in that file.
> 
> Hope that answers your questions.

Thank you very much! Your explanation is very helpful for me.

**Comment by Doragd** [comment]
Hi Kurt, @klshuster, sorry to bother you again.
Are the dstc7 results of the bi-encoder (R@1/100 dev: 56.5 test: 66.8) obtained by training on the augmented data?
If so, I would like to ask about the results of the bi-encoder when training on the original dstc7 data.
